Test that empty trees are properly renamed and then persisted

daniel-wer opened the pull request:
In this test, the whole frontend model is initialized and the sagas are running. This way it's pretty easy to test the tree renaming :) Could be useful for testing other frontend functionality.

### Steps to test:
- No need, the CI running the tests successfully is enough

### Issues:
- avoids the re-occurrence of #2686

------
- [x] Ready for review

philippotto commented on app/assets/javascripts/test/fixtures/tasktracing_server_objects.js:
```diff
@@ -0,0 +1,94 @@
+export const tracing = {
```
Can you add type annotations to the exports so that we enforce that these fixtures are always in sync with our type definitions? I'm not sure how much flow typing is getting lost in the test files due to mocking & co.

philippotto commented on app/assets/javascripts/test/sagas/saga_integration.spec.js:
```diff
@@ -0,0 +1,42 @@
+    name: { $set: generateTreeName(state, treeWithEmptyName.timestamp, treeWithEmptyName.treeId) },
+  });
+
+  const saveQueue = createSaveQueueFromUpdateActions(
```
Can you rename this to `expectedSaveQueue`? I was a bit confused at first why this is manually defined :)

philippotto commented on app/assets/javascripts/test/sagas/saga_integration.spec.js:
```diff
@@ -0,0 +1,42 @@
+
+  // Once the updateTree update action is in the save queue, we're good.
+  // This means the setTreeName action was dispatched, the diffing ran, and the change will be persisted.
+  t.deepEqual(saveQueue, Store.getState().save.queue);
```
You can use `state` here probably?

daniel-wer commented:
@fm3 Flow will fail after my latest commit as I fixed our flow type definitions for some server objects and there was a bug I discovered this way, which I couldn't fix in a simple way. Some time ago, the team property was removed from tasks as it already was determined by the project.
But there is code like this in the frontend:
```javascript
const isAdmin =
      this.props.activeUser.isAdmin ||
      this.props.activeUser.teams
        .filter(team => team.isTeamManager)
        .map(team => team.name)
        .includes(task.team);
```

which is broken right now, as there is no `task.team`.
This code determines whether to show actions like the Task-Transferring for admins (and Team-Managers). The best way to fix this would probably be to request the projects as well and do the cross-check this way, right?

fm3 commented:
@daniel-wer we can also re-add the `team` field to the task json in the backend if you still use it, we can easily look it up in the db via the task’s project when writing the json.

philippotto commented:
Yes, I think that's a good idea since it's easier to look up the team field
in the backend right now.

Florian M <notifications@github.com> schrieb am Fr., 15. Juni 2018, 16:29:

> @daniel-wer <https://github.com/daniel-wer> we can also re-add the team
> field to the task json in the backend if you still use it, we can easily
> look it up in the db via the task’s project when writing the json.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/scalableminds/webknossos/pull/2743#issuecomment-397638904>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACXxGTMq0VG8MVDtVfsyHS_pMg4UTQ-kks5t88TGgaJpZM4UmLt8>
> .
>

daniel-wer commented:
@philippotto I added flow types to our server object fixtures and discovered quite some discrepancies (good call from your side to tell me to do that :) ). I've fixed our flow types and the fixtures as applicable. Please review my last two commits and adjust the flow type `APITaskType` (add team string property) after it was added in the backend. If everything's alright then, feel free to merge while I'm out of office :)

fm3 commented:
@philippotto I added the team name to the task json, please test if the functionality that used it works again.

philippotto commented on app/assets/javascripts/dashboard/explorative_annotations_view.js:
```diff
@@ -357,7 +357,7 @@ class ExplorativeAnnotationsView extends React.PureComponent<Props, State> {
           title="Stats"
           render={(__, tracing: APIAnnotationType) =>
-            tracing.stats && tracing.content.typ === "skeleton" ? (
+            tracing.stats.treeCount && tracing.content.typ === "skeleton" ? (
```
Actually, I wanted to change this to `treeCount != null`, since this won't display the statistics when the tree count is zero. However, this does not typecheck (I opened an issue in the flow repo) **and** the stats object is empty anway, when there aren't any trees (and thus no nodes or edges anyway).

However, I think, the cleanest solution would be that the server always delivers a `stats` object for skeleton tracings and `null` for volume tracings. Even for empty tracings, `stats` should contain attributes (which are `0` then). @fm3 Would this be an easy change? It's not high-pri und doesn't need to block this PR. But if it's easy (and if it makes sense to you), you can change that at some point?

fm3 replied:
not super easy since the stats object is just any json object to the wK backend. it is written only upon update actions via the datastore. but of course we can somehow add a default json object there.

philippotto replied:
ok, I see. then, let's leave it as is :+1:
